snap/channel.go

```diff
@@ -0,0 +1,154 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package snap
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/snapcore/snapd/arch"
+	"github.com/snapcore/snapd/strutil"
+)
+
+var channelRisks = []string{"stable", "candidate", "beta", "edge"}
+
+// StoreChannel identifies and describes completely a store channel.
+type StoreChannel struct {
+	Architecture string `json:"architecture"`
+	Name         string `json:"name"`
+	Track        string `json:"track"`
+	Risk         string `json:"risk"`
+	Branch       string `json:"branch,omitempty"`
+}
+
+// ParseStoreChannel parses a string representing a store channel and includes the given architecture, if architecture is "" the system architecture is included.
+func ParseStoreChannel(s string, architecture string) (StoreChannel, error) {
+	if s == "" {
+		return StoreChannel{}, fmt.Errorf("channel name cannot be empty")
+	}
+	p := strings.Split(s, "/")
+	var risk, track, branch string
+	switch len(p) {
+	default:
+		return StoreChannel{}, fmt.Errorf("channel name has too many components: %s", s)
+	case 3:
+		track, risk, branch = p[0], p[1], p[2]
+	case 2:
+		if strutil.ListContains(channelRisks, p[0]) {
+			risk, branch = p[0], p[1]
+		} else {
+			track, risk = p[0], p[1]
```

Comment:
Is is possible to do some validation of track/branch or are these basically freeform?

Reply:
we could but essentially mostly the store cares, we still be able to produce reasonable errors, actually this is already tying us to things for which mostly only  the store should care at it is